external-crates/move/crates/move-compiler/src/unit_test/plan_builder.rs

```diff
@@ -48,12 +50,31 @@ impl<'env> Context<'env> {
                 (constant.loc, v_opt, constant.attributes.clone())
             })
         });
+        let warning_filters_scope = compilation_env.top_level_warning_filter_scope().clone();
         Self {
             env: compilation_env,
+            warning_filters_scope,
             constants,
         }
     }
 
+    pub fn add_diag(&self, diag: Diagnostic) {
+        self.env.add_diag(&self.warning_filters_scope, diag);
+    }
+
+    #[allow(unused)]
+    pub fn add_diags(&self, diags: Diagnostics) {
+        self.env.add_diags(&self.warning_filters_scope, diags);
+    }
+
+    pub fn add_warning_filter_scope(&mut self, filters: WarningFilters) {
+        self.warning_filters_scope.push(filters)
+    }
+
+    pub fn pop_warning_filter_scope(&mut self) {
+        self.warning_filters_scope.pop()
+    }
+
```

Comment:
Instead of repeating this pattern several places, what about something like:

```
struct DiagnosticReporter {
  diags: Arc<RwLock<Diagnostics>>,
  warning_filters: ...
}

impl DiagnosticReporter {
    pub fn add_diag(&self, diag: Diagnostic) {
        self.env.add_diag(&self.warning_filters_scope, diag);
    }

    #[allow(unused)]
    pub fn add_diags(&self, diags: Diagnostics) {
        self.env.add_diags(&self.warning_filters_scope, diags);
    }

    pub fn add_warning_filter_scope(&mut self, filters: WarningFilters) {
        self.warning_filters_scope.push(filters)
    }

    pub fn pop_warning_filter_scope(&mut self) {
        self.warning_filters_scope.pop()
    }
}
```

And tne for the env:

```
impl CompilationEnv {
  fn diag_reporter() -> { 
    DiagnosticReporter {
      diags: Arc::clone(self.diags),
      warning_filters: self.top_level_warning_filter_scope().clone(),
    }
  }
}
```

Then we can hold/reuse this in local scopes, easing the pressure of the rewrite and simplifying some of the changes.

Reply:
Let's make some progress on this offline. I think it would make this large PR a bit too large. 
I'm fine making iterative progress here, even if eventually this entire PR is undone